mqtt/mqtt-bridge/src/persist/publication_store.rs

```diff
@@ -1,77 +1,87 @@
-use std::sync::Arc;
+use std::{num::NonZeroUsize, sync::Arc};
 
-use anyhow::Result;
 use mqtt3::proto::Publication;
 use parking_lot::Mutex;
 use tracing::debug;
 
-use crate::persist::{
-    loader::MessageLoader, waking_state::StreamWakeableState, Key, PersistError, WakingMemoryStore,
+use crate::{
+    persist::{
+        loader::MessageLoader,
+        waking_state::{memory::WakingMemoryStore, ring_buffer::RingBuffer, StreamWakeableState},
+        Key, PersistResult,
+    },
+    settings::{MemorySettings, RingBufferSettings},
 };
 
 /// Pattern allows for the wrapping `PublicationStore` to be cloned and have non mutable methods
 /// This facilitates sharing between multiple futures in a single threaded environment
 struct PublicationStoreInner<S> {
     state: Arc<Mutex<S>>,
-    offset: u64,
     loader: MessageLoader<S>,
 }
 /// Persistence implementation used for the bridge
 pub struct PublicationStore<S>(Arc<Mutex<PublicationStoreInner<S>>>);
 
 impl PublicationStore<WakingMemoryStore> {
-    pub fn new_memory(batch_size: usize) -> PublicationStore<WakingMemoryStore> {
-        Self::new(WakingMemoryStore::default(), batch_size)
+    pub fn new_memory(
+        batch_size: NonZeroUsize,
+        memory_settings: &MemorySettings,
+    ) -> PublicationStore<WakingMemoryStore> {
+        let max_size = memory_settings.max_size();
+        Self::new(WakingMemoryStore::new(max_size), batch_size)
+    }
+}
+
+impl PublicationStore<RingBuffer> {
+    pub fn new_ring_buffer(
+        batch_size: NonZeroUsize,
+        ring_buffer_settings: &RingBufferSettings,
+        device_id: String,
+        suffix: &str,
+    ) -> PersistResult<Self> {
+        let mut file_path = ring_buffer_settings.directory().clone();
+        file_path.push(device_id);
+        file_path.push(suffix);
```

Comment:
I guess I've missed that before, but why we use device_id and what's the purpose of suffix?

Reply:
Let me add comments here, the purpose is to give a unique qualifier to the path. This is needed because we don't want both local and remote pumps to use the same file and potentially make a mess of things. The device id is nice for isolation both in tests and should someone decide they want to change to a different name.

Reply:
I also had to add this in for this pr to support the default of RingBuffer. Otherwise this would be in a subsequent pr. Once I added in taking values from settings, we default to RingBuffer and then needed to start supporting both memory and RingBuffer. Let me know if these extra files make this all too much, and I will try to come up with something :)